man/man8/xfs_mdrestore.8
.TH xfs_mdrestore 8
.SH NAME
xfs_mdrestore \- restores an XFS metadump image to a filesystem image
.SH SYNOPSIS
.B xfs_mdrestore
[
.B \-gi
] [
.B \-l logdev
] [
.B \-R rtdev
]
.I source
.I target
.br
.B xfs_mdrestore
.B \-i
[
.B \-l logdev
] [
.B \-R rtdev
]
.I source
.br
.B xfs_mdrestore \-V
.SH DESCRIPTION
.B xfs_mdrestore
is a debugging tool that restores a metadata image generated by
.BR xfs_metadump (8)
to a filesystem. The
.I source
argument specifies the location of the metadump image and the
.I target
argument specifies the destination for the filesystem image.
If the
.I source
is \-, then the metadata image is read from stdin. This allows the output of
be another program such as a compression application to be redirected to
.BR xfs_mdrestore .
The
.I target
can be either a file or a device.
.PP
.B xfs_mdrestore
should not be used to restore metadata onto an existing filesystem unless
you are completely certain the
.I target
can be destroyed.
.PP
.SH OPTIONS
.TP
.B \-g
Shows restore progress on stdout.
.TP
.B \-l
Restore log contents to this external log device.
.TP
.B \-i
Shows metadump information on stdout.  If no
.I target
is specified, exits after displaying information.  Older metadumps man not
include any descriptive information.
.TP
.B \-R
Restore realtime device metadata to this device.
.TP
.B \-V
Prints the version number and exits.
.SH DIAGNOSTICS
.B xfs_mdrestore
returns an exit code of 0 if all the metadata is successfully restored or
1 if an error occurs.
.SH SEE ALSO
.BR xfs_metadump (8),
.BR xfs_repair (8),
.BR xfs (5)
.SH BUGS
Email bug reports to
.BR linux-xfs@vger.kernel.org .


debian/changelog
xfsprogs (6.1.0-rc0~WIP-2022-10-14-1djwong1) jammy-experimental; urgency=low

  * Base off for-next on 2022-08-30 (5.19.0+)
  * Support bmbt repairs in xfs_repair.
  * Support realtime rmap btree.
  * Metadata directories.
  * Support atomic file range swaps.
  * AG status and noalloc.
  * Realtime reflink.
  * Better balanced repair workers for scrub.
  * Scrub repair dependencies.
  * reflink/rmapbt/finobt/metadir upgrades.
  * Vectorized scrub experiment.
  * Support for cowextsize on rtreflink files.
  * GETFSREFCOUNTS ioctl and xfs_io utility.
  * Static checking is clean enough (2021-07-15).
  * Prototype free space defragmentation.
  * In-memory reverse mapping btrees for repair.
  * Atomic symlink swapext.
  * Sample mkfs configuration files.
  * Faster xfs_scrub phase5.
  * Scrub file data once per month without a bunch of competing timers.
  * Larger minimum log size and drop support for small filesystems.
  * Multiarch support for libhandle.
  * Better security for xfs_scrub background services.
  * Improve xfs_scrub_all resiliency when systemd gets re-exec'd.
  * Realtime allocation groups.
  * Fixed CPU usage resource control for xfs_scrub processes.
  * Intent items take perag/rtgroup references.

 -- Darrick J. Wong <djwong@djwong.org>  Fri, 14 Oct 2022 09:58:00 -0700

xfsprogs (5.19.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 12 Aug 2022 13:40:25 -0500

xfsprogs (5.18.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 03 Jun 2022 15:18:30 -0400

xfsprogs (5.16.0-1) unstable; urgency=medium

  * New upstream release

  [ Bastian Germann ]
  * debian: Generate .gitcensus instead of .census (Closes: #999743)

 -- Nathan Scott <nathans@debian.org>  Wed, 04 May 2022 14:50:48 -0400

xfsprogs (5.15.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Wed, 06 Apr 2022 16:31:00 -0400

xfsprogs (5.14.2-1) unstable; urgency=medium

  * New upstream release
  * Move rogue fallthrough macro out of linux.h (Closes: #1000974)

 -- Nathan Scott <nathans@debian.org>  Mon, 06 Dec 2021 14:26:57 -0500

xfsprogs (5.14.1-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Thu, 02 Dec 2021 15:26:38 -0500

xfsprogs (5.14.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 19 Nov 2021 10:51:02 -0500

xfsprogs (5.14.0-rc1-1) unstable; urgency=medium

  [ Dave Chinner ]
  * New build dependency: liburcu-dev

  [ Helmut Grohne ]
  * Fix FTCBFS (Closes: #794158)
    + Pass --build and --host to configure

  [ Boian Bonev ]
  * Fix FTBFS (Closes: #997656)
    + Keep custom install-sh after autogen

 -- Bastian Germann <bage@debian.org>  Sun, 14 Nov 2021 23:18:22 +0100

xfsprogs (5.13.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 20 Aug 2021 12:03:31 -0400

xfsprogs (5.12.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 21 May 2021 15:52:00 -0400

xfsprogs (5.11.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 12 Mar 2021 14:08:29 -0500

xfsprogs (5.11.0-rc0-1) experimental; urgency=medium

  [ Steve Langasek ]
  * Regenerate config.guess using debhelper

  [ Bastian Germann ]
  * Build-depend on libinih-dev with udeb package

 -- Bastian Germann <bastiangermann@fishpost.de>  Sat, 20 Feb 2021 11:57:31 +0100

xfsprogs (5.10.0-3) unstable; urgency=medium

  * Drop unused dh-python from Build-Depends (Closes: #981361)
  * Only build for Linux
  * Prevent installing duplicate changelog (Closes: #570704)

 -- Bastian Germann <bastiangermann@fishpost.de>  Fri, 05 Feb 2021 00:18:31 +0100

xfsprogs (5.10.0-2) unstable; urgency=low

  * Team upload
  * debian: cryptographically verify upstream tarball (Closes: #979644)
  * debian: remove dependency on essential util-linux
  * debian: remove "Priority: extra"
  * debian: use Package-Type over its predecessor
  * debian: add missing copyright info (Closes: #979653)

 -- Bastian Germann <bastiangermann@fishpost.de>  Thu, 14 Jan 2021 18:59:14 +0100

xfsprogs (5.10.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 11 Dec 2020 17:16:52 -0500

xfsprogs (5.10.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Fri, 04 Dec 2020 12:19:10 -0500

xfsprogs (5.10.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Mon, 16 Nov 2020 14:26:14 -0600

xfsprogs (5.9.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Tue, 20 Oct 2020 11:38:49 -0400

xfsprogs (5.9.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Tue, 13 Oct 2020 12:18:43 -0400

xfsprogs (5.9.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Thu, 17 Sep 2020 09:11:13 -0500

xfsprogs (5.8.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 04 Sep 2020 15:40:01 -0400

xfsprogs (5.8.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Wed, 26 Aug 2020 14:06:48 -0500

xfsprogs (5.8.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Mon, 10 Aug 2020 14:42:12 -0500

xfsprogs (5.7.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 24 Jul 2020 11:30:00 -0700

xfsprogs (5.7.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Wed, 15 Jul 2020 13:05:24 -0700

xfsprogs (5.7.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Thu, 07 May 2020 13:06:26 -0500


mdrestore/xfs_mdrestore.c
// SPDX-License-Identifier: GPL-2.0
/*
 * Copyright (c) 2007 Silicon Graphics, Inc.
 * All Rights Reserved.
 */

#include "libxfs.h"
#include "xfs_metadump.h"

static int	show_progress = 0;
static int	show_info = 0;
static int	progress_since_warning = 0;

static void
fatal(const char *msg, ...)
{
	va_list		args;

	va_start(args, msg);
	fprintf(stderr, "%s: ", progname);
	vfprintf(stderr, msg, args);
	exit(1);
}

static void
print_progress(const char *fmt, ...)
{
	char		buf[60];
	va_list		ap;

	va_start(ap, fmt);
	vsnprintf(buf, sizeof(buf), fmt, ap);
	va_end(ap);
	buf[sizeof(buf)-1] = '\0';

	printf("\r%-59s", buf);
	fflush(stdout);
	progress_since_warning = 1;
}

extern int	platform_check_ismounted(char *, char *, struct stat *, int);

static int
open_device(
	char		*path,
	int		*is_target_file)
{
	struct stat	statbuf;
	int		open_flags = O_RDWR;
	int		dst_fd;

	*is_target_file = 0;
	if (stat(path, &statbuf) < 0)  {
		/* ok, assume it's a file and create it */
		open_flags |= O_CREAT;
		*is_target_file = 1;
	} else if (S_ISREG(statbuf.st_mode))  {
		open_flags |= O_TRUNC;
		*is_target_file = 1;
	} else  {
		/*
		 * check to make sure a filesystem isn't mounted on the device
		 */
		if (platform_check_ismounted(path, NULL, &statbuf, 0))
			fatal("a filesystem is mounted on target device \"%s\","
				" cannot restore to a mounted filesystem.\n",
				path);
	}

	dst_fd = open(path, open_flags, 0644);
	if (dst_fd < 0)
		fatal("couldn't open target \"%s\"\n", path);

	return dst_fd;
}

static void
check_dev(
	int			dst_fd,
	int			is_target_file,
	unsigned long long	bytes)
{
	if (is_target_file)  {
		/* ensure regular files are correctly sized */

		if (ftruncate(dst_fd, bytes))
			fatal("cannot set filesystem image size: %s\n",
				strerror(errno));
	} else  {
		/* ensure device is sufficiently large enough */

		char		*lb[XFS_MAX_SECTORSIZE] = { NULL };
		off64_t		off;

		off = bytes - sizeof(lb);
		if (pwrite(dst_fd, lb, sizeof(lb), off) < 0)
			fatal("failed to write last block, is target too "
				"small? (error: %s)\n", strerror(errno));
	}
}

/*
 * perform_restore() -- do the actual work to restore the metadump
 *
 * @src_f: A FILE pointer to the source metadump
 * @dst_fd: the file descriptor for the target file
 * @is_target_file: designates whether the target is a regular file
 * @mbp: pointer to metadump's first xfs_metablock, read and verified by the caller
 *
 * src_f should be positioned just past a read the previously validated metablock
 */
static void
perform_restore(
	FILE			*src_f,
	int			dst_fd,
	int			is_target_file,
	const struct xfs_metablock	*mbp,
	char			*log_path)
{
	struct xfs_metablock	*metablock;	/* header + index + blocks */
	__be64			*block_index;
	char			*block_buffer;
	int			block_size;
	int			max_indices;
	int			cur_index;
	int			mb_count;
	xfs_sb_t		sb;
	int64_t			bytes_read;
	int64_t			mb_read = 0;
	int			log_fd = -1;
	bool			is_mdx;

	is_mdx = mbp->mb_magic == cpu_to_be32(XFS_MDX_MAGIC);

	block_size = 1 << mbp->mb_blocklog;
	max_indices = (block_size - sizeof(xfs_metablock_t)) / sizeof(__be64);

	metablock = (xfs_metablock_t *)calloc(max_indices + 1, block_size);
	if (metablock == NULL)
		fatal("memory allocation failure\n");

	mb_count = be16_to_cpu(mbp->mb_count);
	if (mb_count == 0 || mb_count > max_indices)
		fatal("bad block count: %u\n", mb_count);

	memcpy(metablock, mbp, sizeof(struct xfs_metablock));
	block_index = (__be64 *)((char *)metablock + sizeof(xfs_metablock_t));
	block_buffer = (char *)metablock + block_size;

	if (fread(block_index, block_size - sizeof(struct xfs_metablock), 1, src_f) != 1)
		fatal("error reading from metadump file\n");

	if (block_index[0] != 0)
		fatal("first block is not the primary superblock\n");


	if (fread(block_buffer, mb_count << mbp->mb_blocklog, 1, src_f) != 1)
		fatal("error reading from metadump file\n");

	libxfs_sb_from_disk(&sb, (struct xfs_dsb *)block_buffer);

	if (sb.sb_magicnum != XFS_SB_MAGIC)
		fatal("bad magic number for primary superblock\n");

	/*
	 * Normally the upper bound would be simply XFS_MAX_SECTORSIZE
	 * but the metadump format has a maximum number of BBSIZE blocks
	 * it can store in a single metablock.
	 */
	if (sb.sb_sectsize < XFS_MIN_SECTORSIZE ||
	    sb.sb_sectsize > XFS_MAX_SECTORSIZE ||
	    sb.sb_sectsize > max_indices * block_size)
		fatal("bad sector size %u in metadump image\n", sb.sb_sectsize);

	((struct xfs_dsb*)block_buffer)->sb_inprogress = 1;

	check_dev(dst_fd, is_target_file, sb.sb_dblocks * sb.sb_blocksize);

	bytes_read = 0;

	for (;;) {
		int		write_fd = dst_fd;

		if (metablock->mb_magic != mbp->mb_magic)
			fatal("magic value wrong");

		if (metablock->mb_info & XFS_METADUMP_LOGDEV) {
			int	log_is_file;

			if (!is_mdx)
				fatal("logdev set on an old style metadump?");
			if (log_fd == -1) {
				if (!log_path)
					fatal(
	"metadump has log contents but -l was not specified?");
				log_fd = open_device(log_path, &log_is_file);
				check_dev(log_fd, log_is_file,
						sb.sb_logblocks * sb.sb_blocksize);
			}
			write_fd = log_fd;
		}
		if (metablock->mb_info & XFS_METADUMP_RTDEV) {
			if (!is_mdx)
				fatal("rtdev set on an old style metadump?");
			fatal("rtdev not supported");
		}

		if (show_progress) {
			int64_t		mb_now = bytes_read >> 20;

			if (mb_now != mb_read) {
				print_progress("%lld MB read", mb_now);
				mb_read = mb_now;
			}
		}

		for (cur_index = 0; cur_index < mb_count; cur_index++) {
			if (pwrite(write_fd, &block_buffer[cur_index <<
					mbp->mb_blocklog], block_size,
					be64_to_cpu(block_index[cur_index]) <<
						BBSHIFT) < 0)
				fatal("error writing block %llu: %s\n",
					be64_to_cpu(block_index[cur_index]) << BBSHIFT,
					strerror(errno));
		}
		if (is_mdx) {
			size_t	nr = fread(metablock, block_size, 1, src_f);

			if (nr == 0)
				break;
			if (nr != 1)
				fatal("error reading from extended metadump file\n");
		} else {
			if (mb_count < max_indices)
				break;

			if (fread(metablock, block_size, 1, src_f) != 1)
				fatal("error reading from metadump file\n");
		}

		mb_count = be16_to_cpu(metablock->mb_count);
		if (mb_count == 0)
			break;
		if (mb_count > max_indices)
			fatal("bad block count: %u\n", mb_count);

		if (fread(block_buffer, mb_count << mbp->mb_blocklog,
								1, src_f) != 1)
			fatal("error reading from metadump file\n");

		bytes_read += block_size + (mb_count << mbp->mb_blocklog);
	}

	if (show_progress && bytes_read > (mb_read << 20))
		print_progress("%lld MB read", mb_read + 1);

	if (progress_since_warning)
		putchar('\n');

	memset(block_buffer, 0, sb.sb_sectsize);
	sb.sb_inprogress = 0;
	libxfs_sb_to_disk((struct xfs_dsb *)block_buffer, &sb);
	if (xfs_sb_version_hascrc(&sb))
		xfs_update_cksum(block_buffer, sb.sb_sectsize, XFS_SB_CRC_OFF);

	if (pwrite(dst_fd, block_buffer, sb.sb_sectsize, 0) < 0)
		fatal("error writing primary superblock: %s\n", strerror(errno));

	if (log_fd >= 0)
		close(log_fd);

	free(metablock);
}

static void
usage(void)
{
	fprintf(stderr, "Usage: %s [-V] [-g] [-i] [-l logdev] source target\n", progname);
	exit(1);
}

int
main(
	int 		argc,
	char 		**argv)
{
	char		*log_path = NULL;
	FILE		*src_f;
	int		dst_fd;
	int		c;
	int		is_target_file;
	struct xfs_metablock	mb;

	progname = basename(argv[0]);

	while ((c = getopt(argc, argv, "gl:iV")) != EOF) {
		switch (c) {
			case 'g':
				show_progress = 1;
				break;
			case 'l':
				log_path = optarg;
				break;
			case 'i':
				show_info = 1;
				break;
			case 'V':
				printf("%s version %s\n", progname, VERSION);
				exit(0);
			default:
				usage();
		}
	}

	if (argc - optind < 1 || argc - optind > 2)
		usage();

	/* show_info without a target is ok */
	if (!show_info && argc - optind != 2)
		usage();

	/*
	 * open source and test if this really is a dump. The first metadump block
	 * will be passed to perform_restore() which will continue to read the
	 * file from this point. This avoids rewind the stream, which causes
	 * restore to fail when source was being read from stdin.
 	 */
	if (strcmp(argv[optind], "-") == 0) {
		src_f = stdin;
		if (isatty(fileno(stdin)))
			fatal("cannot read from a terminal\n");
	} else {
		src_f = fopen(argv[optind], "rb");
		if (src_f == NULL)
			fatal("cannot open source dump file\n");
	}

	if (fread(&mb, sizeof(mb), 1, src_f) != 1)
		fatal("error reading from metadump file\n");
	if (mb.mb_magic != cpu_to_be32(XFS_MD_MAGIC) &&
	    mb.mb_magic != cpu_to_be32(XFS_MDX_MAGIC))
		fatal("specified file is not a metadata dump\n");

	if (show_info) {
		if (mb.mb_info & XFS_METADUMP_INFO_FLAGS) {
			printf("%s: %sobfuscated, %s log, %s metadata blocks\n",
			argv[optind],
			mb.mb_info & XFS_METADUMP_OBFUSCATED ? "":"not ",
			mb.mb_info & XFS_METADUMP_DIRTYLOG ? "dirty":"clean",
			mb.mb_info & XFS_METADUMP_FULLBLOCKS ? "full":"zeroed");
		} else {
			printf("%s: no informational flags present\n",
				argv[optind]);
		}

		if (argc - optind == 1)
			exit(0);
	}

	optind++;

	/* check and open target */
	dst_fd = open_device(argv[optind], &is_target_file);

	perform_restore(src_f, dst_fd, is_target_file, &mb, log_path);

	close(dst_fd);
	if (src_f != stdin)
		fclose(src_f);

	return 0;
}
